src/dmd/cppmangle.d

```diff
@@ -584,7 +584,20 @@ private final class CppMangleVisitor : Visitor
              */
             TemplateInstance ti = d.parent.isTemplateInstance();
             assert(ti);
-            source_name(ti);
+            Dsymbol p = ti.toParent();
+            if (p && !p.isModule() && tf.linkage == LINK.cpp)
+            {
+                buf.writeByte('N');
+                CV_qualifiers(d.type);
+                prefix_name(p);
+                if (d.isDtorDeclaration())
+                    buf.writestring("D1");
+                else
+                    source_name(ti);
+                buf.writeByte('E');
+            }
+            else
+                source_name(ti);
```

Comment:
This just seems to be a duplicate of below, maybe it could be refactored in such a way that they are made a bit more common?

Perhaps it could be benefited from splitting into a new function.

Reply:
It's not quite a duplicate. I also wanted to make this easy to review, and including a refactoring of other code would make it harder.